rust/src/passwd.rs

```diff
@@ -1,17 +1,146 @@
+/// /usr/lib/passwd, so that %pre scripts are aware of newly added system users
+/// not in the tree's /usr/lib/passwd (through nss-altfiles in the container).
+pub fn prepare_rpm_layering(rootfs_dfd: i32, merge_passwd_dir: &str) -> anyhow::Result<bool> {
+    passwd_cleanup(rootfs_dfd)?;
+    let rootfs = ffiutil::ffi_view_openat_dir(rootfs_dfd);
+    let dir: Option<PathBuf> = if merge_passwd_dir.is_empty() {
+        None
+    } else {
+        Some(merge_passwd_dir.into())
+    };
+
+    // Break hardlinks for the shadow files, since shadow-utils currently uses
+    // O_RDWR unconditionally.
+    for filename in PWGRP_SHADOW_FILES {
+        let src = format!("etc/{}", filename);
+        if rootfs.exists(&src)? {
+            ostree::break_hardlink(
+                rootfs.as_raw_fd(),
+                &src,
+                true,
+                Option::<&gio::Cancellable>::None,
+            )?;
+        };
+    }
+
+    let has_usrlib_passwd = has_usrlib_passwd(&rootfs)?;
+    if has_usrlib_passwd {
+        prepare_pwgrp(&rootfs, dir)?;
+    }
+
+    Ok(has_usrlib_passwd)
+}
+
+pub fn complete_rpm_layering(rootfs_dfd: i32) -> anyhow::Result<()> {
+    let rootfs = ffiutil::ffi_view_openat_dir(rootfs_dfd);
+    complete_pwgrp(&rootfs)?;
+
+    Ok(())
+}
+
+/// Clean up passwd files.
+///
+/// This may be leftover in the tree from an older version of rpm-ostree that
+/// didn't clean them up at compose time, and having them exist will mean
+/// rofiles-fuse will prevent useradd from opening it for write.
+pub fn passwd_cleanup(rootfs_dfd: i32) -> anyhow::Result<()> {
+    let rootfs = ffiutil::ffi_view_openat_dir(rootfs_dfd);
+    for filename in PWGRP_LOCK_AND_BACKUP_FILES {
+        let target = format!("usr/etc/{}", filename);
+        rootfs.remove_file_optional(target)?;
+    }
+
+    Ok(())
+}
+
+fn has_usrlib_passwd(rootfs: &openat::Dir) -> anyhow::Result<bool> {
+    // Does this rootfs have a usr/lib/passwd? We might be doing a
+    // container or something else.
+    let has_passwd = rootfs.exists("usr/lib/passwd")?;
+    Ok(has_passwd)
+}
+
+fn prepare_pwgrp(rootfs: &openat::Dir, merge_passwd_dir: Option<PathBuf>) -> anyhow::Result<()> {
+    for filename in USRLIB_PWGRP_FILES {
+        let etc_file = format!("etc/{}", filename);
+        let etc_backup = format!("{}.rpmostreesave", etc_file);
+        let usrlib_file = format!("usr/lib/{}", filename);
+        let usrlib_file_tmp = format!("{}.tmp", &usrlib_file);
+
+        // Retain the current copies in /etc as backups.
+        rootfs.local_rename(&etc_file, &etc_backup)?;
+
+        // Copy /usr/lib/{passwd,group} -> /etc (breaking hardlinks).
+        utils::file_copy_noxattr(rootfs, &usrlib_file, rootfs, &etc_file, false)?;
+
+        // Copy the merge's passwd/group to usr/lib (breaking hardlinks).
+        if let Some(ref merge_dir) = merge_passwd_dir {
+            let cwd = openat::Dir::open(".")?;
```

Comment:
Hmm, odd that the original code has this relative to the process working directory.  Ah I see, the path is always absolute.  So we could perhaps just use `std::fs::copy(merge_file, format!("/proc/self/fd/{}", rootfs.as_raw_fd())` or so.

It's kind of annoying that the `openat` crate doesn't bind `AT_FDCWD`.

Reply:
I'm not sure that `fs::copy` has the same semantics (regarding overwrites and xattrs) of the glnx helper.
Would it maybe make more sense to change this into `Dir::open("/")` here to improve readability, and later switch to an `openat-ext` port of the glnx helper?

Reply:
OK I put up https://github.com/cgwalters/openat-ext/pull/16

Reply:
Thanks, reviewed.
However that only covers the `overwrite: true` case, which isn't enough to cover all the cases in this PR.

Going to back to the same unanswered question in https://github.com/coreos/rpm-ostree/pull/2362#discussion_r556703651, do you plan to have an additional `copy_file_optional()`? Or how do you prefer passing/encoding option flags in that library (there is also a potential `COPY_NOXATTRS` at play).

Reply:
Hmm the cases where one wants to error if the target exists are actually pretty obscure IMO.  Are there cases here where you think that's a "must have"?

I think any case where one wants to error on target existing are actually much more simply implemented by using https://docs.rs/openat/0.1.19/openat/struct.Dir.html#method.new_file
(And if one really cares, add code to delete the file on error)

Reply:
There is one such case in this code in `prepare_pwgrp()`. Though on a second look the surrounding code there guarantees that the destination does not exist (it is renamed for backup just before).
So indeed we should be clear with dropping all C compat with the new openat-ext release.

Reply:
I opened https://github.com/cgwalters/openat-ext/pull/17 which possibly helps in taking care of this case (source dirfd != target dirfd) without going through /proc.